### ISSUE TITLE
Bug Report: HwAMC::getTriggerLinkStatus Doesn't Catch CSC Trigger Fiber is Unplugged

### ISSUE THREAD

**Issue by bdorney** [cms-gem-daq-project/cmsgemos#268]
## Brief summary of issue

Prep'ing for mass production and I noticed that the detector on the QC7 setup doesn't have the CSC Trigger fiber plugged in and `getTriggerLinkStatus` reports all counters as `0x0` and instead of returning true or false it returns the sum of all counters. 

### Types of issue

- [x] Bug report (report an issue with the code)
- [ ] Feature request (request for change which adds functionality)

## Expected Behavior

Should accurately report the trigger link status.

## Current Behavior

```python
In [1]: from gempython.tools.amc_user_functions_xhal import HwAMC

In [2]: amcBoard = HwAMC(cardName="gem-shelf02-amc05")
Open pickled address table if available  /opt/cmsgemos/etc/maps//amc_address_table_top.pickle...
Initializing AMC gem-shelf02-amc05

In [3]: amcBoard.getTriggerLinkStatus(printSummary=True,checkCSCTrigLink=True,ohMask=0x10)
--=======================================--
-> GEM SYSTEM TRIGGER LINK INFORMATION
--=======================================--

----------OH4----------
LINK0_MISSED_COMMA_CNT		0
LINK1_MISSED_COMMA_CNT		0
LINK0_OVERFLOW_CNT		0
LINK1_OVERFLOW_CNT		0
LINK0_UNDERFLOW_CNT		0
LINK1_UNDERFLOW_CNT		0
LINK0_SBIT_OVERFLOW_CNT		0
LINK1_SBIT_OVERFLOW_CNT		0
----------OH5----------
LINK0_MISSED_COMMA_CNT		0
LINK1_MISSED_COMMA_CNT		0
LINK0_OVERFLOW_CNT		0
LINK1_OVERFLOW_CNT		0
LINK0_UNDERFLOW_CNT		0
LINK1_UNDERFLOW_CNT		0
LINK0_SBIT_OVERFLOW_CNT		0
LINK1_SBIT_OVERFLOW_CNT		0
Out[3]: 0L
```

While:

```bash
$ gem_reg.py
Open pickled address table if available  /opt/cmsgemos/etc/maps//amc_address_table_top.pickle...
Starting CTP7 Register Command Line Interface. Please connect to CTP7 using connect <hostname> command unless you use it directly at the CTP7
CTP7 > connect eagle60
eagle60 > kw GEM_AMC.TRIGGER.OH4
...
...
0x66001680 r 	GEM_AMC.TRIGGER.OH4.LINK0_SBIT_OVERFLOW_CNT		0x00000000
0x66001680 r 	GEM_AMC.TRIGGER.OH4.LINK1_SBIT_OVERFLOW_CNT		0x00000000
0x66001684 r 	GEM_AMC.TRIGGER.OH4.LINK0_MISSED_COMMA_CNT		0x00000000
0x66001684 r 	GEM_AMC.TRIGGER.OH4.LINK1_MISSED_COMMA_CNT		0x00000000
0x6600168c r 	GEM_AMC.TRIGGER.OH4.LINK0_OVERFLOW_CNT			0x00000000
0x6600168c r 	GEM_AMC.TRIGGER.OH4.LINK1_OVERFLOW_CNT			0x00000000
0x66001690 r 	GEM_AMC.TRIGGER.OH4.LINK0_UNDERFLOW_CNT			0x00000005
0x66001690 r 	GEM_AMC.TRIGGER.OH4.LINK1_UNDERFLOW_CNT			0x00000004
0x66001694 r 	GEM_AMC.TRIGGER.OH4.LINK0_SYNC_WORD_CNT			0x00000000
0x66001694 r 	GEM_AMC.TRIGGER.OH4.LINK1_SYNC_WORD_CNT			0x00000000
eagle60 > kw GEM_AMC.TRIGGER.OH5
...
...
0x66001a80 r 	GEM_AMC.TRIGGER.OH5.LINK0_SBIT_OVERFLOW_CNT		0x0000ffff
0x66001a80 r 	GEM_AMC.TRIGGER.OH5.LINK1_SBIT_OVERFLOW_CNT		0x0000ffff
0x66001a84 r 	GEM_AMC.TRIGGER.OH5.LINK0_MISSED_COMMA_CNT		0x0000ffff
0x66001a84 r 	GEM_AMC.TRIGGER.OH5.LINK1_MISSED_COMMA_CNT		0x0000ffff
0x66001a8c r 	GEM_AMC.TRIGGER.OH5.LINK0_OVERFLOW_CNT			0x00000000
0x66001a8c r 	GEM_AMC.TRIGGER.OH5.LINK1_OVERFLOW_CNT			0x00000000
0x66001a90 r 	GEM_AMC.TRIGGER.OH5.LINK0_UNDERFLOW_CNT			0x0000ffff
0x66001a90 r 	GEM_AMC.TRIGGER.OH5.LINK1_UNDERFLOW_CNT			0x0000ffff
0x66001a94 r 	GEM_AMC.TRIGGER.OH5.LINK0_SYNC_WORD_CNT			0x00000000
0x66001a94 r 	GEM_AMC.TRIGGER.OH5.LINK1_SYNC_WORD_CNT			0x00000000
```

So there's definitely a problem.

### Steps to Reproduce (for bugs)

Shown above.

## Possible Solution (for bugs)

For one I think this is a problem:

https://github.com/cms-gem-daq-project/cmsgemos/blob/bbaeabb01c7ab4c8862d0ad5c0cb6a1cfe973efa/gempython/tools/amc_user_functions_xhal.py#L425

## Context

Doesn't qualify the trigger link status in QC7.

## Your Environment

* Version used: bbaeabb01c7ab4c8862d0ad5c0cb6a1cfe973efa

**Comment by bdorney** [comment]
I'm going to start working on a solution immediately.

**Comment by BenjaminRS** [comment]
This looks very odd -- is it possible that this happened due to squashing?

**Comment by bdorney** [comment]
> This looks very odd -- is it possible that this happened due to squashing?

Yes `git` can do some strange things when merging/squashing.

**Comment by bdorney** [comment]
Okay there where two problems, the first is the assignment error:

https://github.com/cms-gem-daq-project/cmsgemos/blob/bbaeabb01c7ab4c8862d0ad5c0cb6a1cfe973efa/gempython/tools/amc_user_functions_xhal.py#L425

The second is a bad indexing convention here:

https://github.com/cms-gem-daq-project/cmsgemos/blob/bbaeabb01c7ab4c8862d0ad5c0cb6a1cfe973efa/gempython/tools/amc_user_functions_xhal.py#L450-L457

And also here:

https://github.com/cms-gem-daq-project/cmsgemos/blob/bbaeabb01c7ab4c8862d0ad5c0cb6a1cfe973efa/gempython/tools/amc_user_functions_xhal.py#L464-L471

Here the indices are iterating such that each set of 8 indices corresponds to one OH since there are 8 registers per 8.  But in reality:

https://github.com/cms-gem-daq-project/xhal/blob/99dffb6b3cd8a4053b33515d46fb79ced042ecd5/xhalcore/src/common/rpc_manager/daq_monitor.cc#L91-L106

Each set of `self.nOHs` indices corresponds to register X (e.g. `LINK1_UNDERFLOW_CNT`) for a set of  `self.nOHs` number of optohybrids.

So the solution is to change the indexing convention to what is mentioned in `xhal`.

@BenjaminRS I'll apply this tomorrow and prepare a PR.

**Comment by bdorney** [comment]
Yeah I found another bug:

```
(py2.7) [dorney@gem904qc8daq]~/scratch0/CMS_GEM/CMS_GEM_DAQ% testConnectivity.py -d --shelf=2 -s5 -o 0x10 --nPhaseScans=100 --skipDACScan --skipScurve --checkCSCTrigLink
...
...
====================
Step 3: Programming FPGA & Checking FPGA Communication
====================
Initial value to write: 1, register GEM_AMC.TTC.GENERATOR.ENABLE
Initial value to write: 1, register GEM_AMC.TTC.GENERATOR.SINGLE_HARD_RESET
Initial value to write: 1, register GEM_AMC.TTC.GENERATOR.SINGLE_HARD_RESET
Initial value to write: 1, register GEM_AMC.TTC.GENERATOR.SINGLE_HARD_RESET
Initial value to write: 1, register GEM_AMC.TTC.GENERATOR.SINGLE_HARD_RESET
Initial value to write: 1, register GEM_AMC.TTC.GENERATOR.SINGLE_HARD_RESET
Initial value to write: 0, register GEM_AMC.TTC.GENERATOR.ENABLE
Checking trigger link status:
--=======================================--
-> GEM SYSTEM TRIGGER LINK INFORMATION
--=======================================--

----------OH4----------
LINK0_MISSED_COMMA_CNT          0xffff
LINK1_MISSED_COMMA_CNT          0x0
LINK0_OVERFLOW_CNT              0x0
LINK1_OVERFLOW_CNT              0x0
LINK0_UNDERFLOW_CNT             0xffff
LINK1_UNDERFLOW_CNT             0x0
LINK0_SBIT_OVERFLOW_CNT         0xffff
LINK1_SBIT_OVERFLOW_CNT         0x0
----------OH5----------
LINK0_MISSED_COMMA_CNT          0x0
LINK1_MISSED_COMMA_CNT          0x0
LINK0_OVERFLOW_CNT              0x0
LINK1_OVERFLOW_CNT              0x0
LINK0_UNDERFLOW_CNT             0x0
LINK1_UNDERFLOW_CNT             0x0
LINK0_SBIT_OVERFLOW_CNT         0x0
LINK1_SBIT_OVERFLOW_CNT         0x0
Traceback (most recent call last):
  File "/afs/cern.ch/user/d/dorney/scratch0/CMS_GEM/CMS_GEM_DAQ/venv/cc7/py2.7/lib/python2.7/site-packages/gempython/scripts/testConnectivity.py", line 929, in <module>
    testConnectivity(args)
  File "/afs/cern.ch/user/d/dorney/scratch0/CMS_GEM/CMS_GEM_DAQ/venv/cc7/py2.7/lib/python2.7/site-packages/gempython/scripts/testConnectivity.py", line 252, in testConnectivity
    ohMask=args.ohMask)
  File "/afs/cern.ch/user/d/dorney/scratch0/CMS_GEM/CMS_GEM_DAQ/venv/cc7/py2.7/lib/python2.7/site-packages/gempython/tools/amc_user_functions_xhal.py", line 479, in getTriggerLinkStatus
    ohSumLinkStatus[ohN+1]+=linkResult[(ohN+1)+idx*self.nOHs]
KeyError: 5
```

**Comment by BenjaminRS** [comment]
Where did the bug come from? I assume it is when asking for linkResult[(ohN+1)+idx*self.nOHs] ?

**Comment by bdorney** [comment]
linkResult was initialized to all zeros so not the problem; it was ohSumLinkStatus (that I added) that didn’t have the OHN+1 key added when checking CSC trigger link.

**Comment by bdorney** [comment]
Closed by #269
